rfcs/v3/rest_json_api.md

```diff
@@ -0,0 +1,258 @@
+  if (!author) {
+	return "Invalid author";
+  }
+  return author.name;
+}
+```
+
+Type to Command Relationship
+
+```
+kind: Relationship
+version: v1
+definition:
+  name: author_name
+  source: articles
+  target:
+	command:
+  		name: author_name
+  		subgraph: null
+  mapping:
+	- source:
+    		fieldPath:
+      			- fieldName: author
+  	  target:
+    		argument:
+    		  	argumentName: author
+```
+
+In the future when computed columns are supported in OpenDD, it can be part of the general response
+type and not relationships
+
+## Response Document Structure
+
+Responses MUST be according to the JSON:API document structure.
+
+* JSON object must be root of every document
+* A document MUST contain at least one of the following top-level members:
+
+```
+  data: the document’s “primary data”, either a single resource object, an array of resource objects, or null.
+
+  errors: an array of error objects.
+
+  meta: contains non-standard meta-information.
+```
+
+* The members \`data\` and \`errors\` MUST NOT coexist in the same document.
+* Primary data MUST be
+  * single resource object
+  * array of resource objects
+
+### Top-Level Structure:
+
+- A JSON document **MUST** contain at least one of the following top-level members:
+  - `data`: The document’s “primary data”.
+    - is either a single resource object, an array of resource objects, or null.
+  - `errors`: An array of error objects.
+  - `meta`: A meta object that contains non-standard meta-information.
+- Other Top Level Items
+  - `included:` Includes related (to the primary data) resource objects and is used to provide
+    compound documents. The purpose of this section is to reduce the number of requests clients need
+    to make by including related resources directly in the response.
+    - Compound Documents \- Allow responses to include related resources along with the primary
+      resources.
+      - A [compound document](https://jsonapi.org/format/\#document-compound-documents) **MUST NOT**
+        include more than one resource object for each `type` and `id` pair.
+      - Every included resource object **MUST** be identified via a chain of relationships
+        originating in a document’s primary data. This means that compound documents require “full
+        linkage” and that no resource object can be included without a direct or indirect
+        relationship to the document’s primary data.
+    - Can also contain relationships and related links, just like the primary `data` resource
+      objects
+  - `links`: Top-level links object contains pagination links (`self`, `next`, `last`).
+  - `jsonapi:` Contains the version of the JSON
+  - being used and additional meta-information about the API, including a `describedby` URL.
+
+Read more on the [resource object](https://jsonapi.org/format/\#document-resource-objects).
+
+# OpenDD config
+
+For the first iteration, to keep things simple, we are not introducing any OpenDD configuration. Any
```

Comment:
I like this because we're keeping scope slim, but I am concerned about breakages - ie, if by default we expose all models as a JSON:API endpoint, and then we later make each model opt-in, does that constitute a breaking change?

Do we solve this by only offering this feature on a strictly "this is experimental" basis?

Reply:
That's a good point. It will be a breaking change, and I guess we would have to incorporate this via Compatibility Date. But let me see if we can do anything better here.

Reply:
Oh of course. If we just say "everything before this date gets all models, everything after gets specific ones", we're fine.

Reply:
Yeah, I'm thinking we can introduce a date after which it is still all models but user gets a warning. And after a second date, we switch to specific ones. Does this make sense?

Reply:
Yeah, sounds ideal!